### PR TITLE
paging: Support five levels of paging if host supports it

### PR DESCRIPTION
**Pull request by rbradford**
Set up 5 levels of identity pages and enable 5-level paging in CR4.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

### CODE REVIEW

**Comment by josephlr** [comment]
I'm not super familiar with the 5 level paging feature, what's the reason to use the fifth level if we're only mapping the first 4 GiB? Or is there a need to support physical/virtual addresses outside of the standard 4-level range?

**Comment by rbradford** [comment]
@josephlr I think in some cases the Linux kernel will not enable L5 paging if the firmware hasn't already set this up. I'm still checking if this is necessary so this is not something to be merged yet.

**Comment by rbradford** [comment]
(It's also worth noting that OVMF sets up 5 levels of identity page tables and sets CR4 where appropriate.)

File: src/paging.rs [hunk 1]
```diff
@@ -39,13 +42,31 @@ pub fn setup() {
...
+        let (cr3_frame, cr3_flags) = Cr3::read();
+        let l5_frame = PhysFrame::from_start_address(phys_addr(l5)).unwrap();
+        if cr3_frame != l5_frame {
+            unsafe { Cr3::write(l5_frame, cr3_flags) };
+        }
+        let mut cr4 = Cr4::read();
+        cr4.insert(Cr4Flags::L5_PAGING);
+        unsafe {
+            Cr4::write(cr4);
+        }
```
Comment (josephlr) [inline src/paging.rs:58]: Nit: Format like: `unsafe { Cr4::write(cr4) };` for consistency with the `Cr3` code.

File: src/paging.rs [hunk 2]
```diff
@@ -39,13 +42,31 @@ pub fn setup() {
...
-        unsafe { Cr3::write(l4_frame, cr3_flags) };
+    if unsafe { core::arch::x86_64::__cpuid(7).ecx } & (1 << 16) != 0 {
+        // Point L5 at L4
+        l5[0].set_addr(phys_addr(l4), pt_flags);
+        // Point Cr3 at L5
+        let (cr3_frame, cr3_flags) = Cr3::read();
+        let l5_frame = PhysFrame::from_start_address(phys_addr(l5)).unwrap();
+        if cr3_frame != l5_frame {
+            unsafe { Cr3::write(l5_frame, cr3_flags) };
+        }
```
Comment (josephlr) [inline src/paging.rs:53]: If we do this, we should probably unconditionally write `Cr3`. I'm not sure of the exact TLB semantics, but I think if we update the tables at all we need the TLB to be flushed.

Same is true for the `l4_frame` stuff below.

**Comment by josephlr** [comment]
> @josephlr I think in some cases the Linux kernel will not enable L5 paging if the firmware hasn't already set this up. I'm still checking if this is necessary so this is not something to be merged yet.

Looking at Linux Source, it seems like the kernel [supports 4 -> 5 level transition](https://github.com/torvalds/linux/blob/9f159ae07f07fc540290f21937231034f554bdd7/arch/x86/boot/compressed/pgtable_64.c#L110), but I'm not sure if that's hit on all boot paths.

**Comment by rbradford** [comment]
Yeh, we don't need this

> > @josephlr I think in some cases the Linux kernel will not enable L5 paging if the firmware hasn't already set this up. I'm still checking if this is necessary so this is not something to be merged yet.
> 
> Looking at Linux Source, it seems like the kernel [supports 4 -> 5 level transition](https://github.com/torvalds/linux/blob/9f159ae07f07fc540290f21937231034f554bdd7/arch/x86/boot/compressed/pgtable_64.c#L110), but I'm not sure if that's hit on all boot paths.

Yeh, we don't need this.

**Pull request closed without merging**